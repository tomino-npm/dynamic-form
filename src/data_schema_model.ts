import * as validations from './validation';

// @ts-ignore
import initCustomErrors from 'ajv-errors';
import Ajv from 'ajv';

import { getRoot, getPath, IAnyStateTreeNode } from 'mobx-state-tree';

import { JSONSchemaType, JSONSchema } from './json_schema';
import { safeEval } from './form_utils';
import { DataSet } from './form_store';

const defaultAjv = new Ajv({ allErrors: true, useDefaults: true });
// initCustomErrors(defaultAjv);

defaultAjv.addKeyword('validationExpression', {
  // type: 'string',]
  errors: true,
  validate: function v(
    this: any,
    schema: any,
    data: any,
    elementSchema: JSONSchema,
    elementPath: string,
    parentData: any
  ) {
    // console.log(arguments);
    // console.log(parentData);
    // console.log(schema);
    // console.log(data);
    // console.log(safeEval(parentData, schema, data));
    // console.log(elementPath);

    const result = safeEval(parentData, schema, data);
    if (!result) {
      // console.log(`Validation Error "${elementSchema.validationMessage}" at: ` + elementPath);

      (v as any).errors = [];
      (v as any).errors.push({
        keyword: 'validationExpression',
        dataPath: elementPath,
        message: elementSchema.validationMessage
          ? elementSchema.validationMessage
          : 'Value is invalid',
        params: {
          keyword: 'validationExpression'
        }
      });
    }
    return result;
  }
});

export type ListItem = {
  text: string;
  value: string;
};

type SchemaOptions = {
  parent?: Schema;
  required?: boolean;
  key?: string;
  ajv?: Ajv.Ajv;
};

export class Schema {
  // store: typeof FormStore.Type;
  parent: Schema;
  properties: { [index: string]: Schema };
  items: Schema;
  type: JSONSchemaType;
  default: any;
  required: boolean;
  readOnly: boolean;
  validationMessage: string;
  expression: string;
  enum: ListItem[];
  key: string;
  schema: JSONSchema;

  validator: Ajv.ValidateFunction;
  ajv: Ajv.Ajv;

  constructor(
    schema: JSONSchema,
    { parent = null, required = false, key = null, ajv = defaultAjv }: SchemaOptions = {}
  ) {
    this.parent = parent;
    this.schema = schema;
    this.expression = schema.expression;
    this.key = key;

    // we do not need validator for end nodes
    // these are always validated from the parent
    if (schema.properties) {
      this.ajv = ajv;
      this.validator = this.ajv.compile(schema);
    }

    this.init(schema, ['readOnly', 'type', 'required', 'default', 'validationMessage', 'enum']);

    if (required) {
      this.required = required;
    }

    if (schema.type === 'object') {
      this.properties = {};
      for (let key of Object.getOwnPropertyNames(schema.properties)) {
        this.properties[key] = new Schema(schema.properties[key] as JSONSchema, {
          parent: this,
          required: schema.required && schema.required.includes(key),
          key
        });
      }
    }

    if (schema.type === 'array') {
      this.items = new Schema(schema.items as JSONSchema, this);
    }
  }

  rootSchema() {
    let s: Schema = this;
    while (s.parent != null) {
      s = s.parent;
    }
    return s;
  }

  // randomValue() {
  //   switch (this.type) {
  //     case 'string':
  //       return random.words(2);
  //     case 'boolean':
  //       return random.boolean();
  //     case 'id':
  //       return '1';
  //     case 'integer':
  //       return random.int();
  //     case 'number':
  //       return random.float();
  //   }
  //   return undefined;
  // }

  init<T>(schema: T, keys: Array<keyof Schema>) {
    for (let key of keys) {
      if ((schema as any)[key] != null) {
        this[key] = (schema as any)[key];
      }
    }
  }

  defaultValue<T>(dataset: T = {} as T): T {
    // validator is set to assign default values
    this.validator(dataset);
    return dataset;
  }

  /* =========================================================
      Validations
     ======================================================== */

  static convertPath(path: string) {
    path = path.replace(/\/(\d+)/g, '[$1]');
    path = path.replace(/\//g, '.');

    return path;
  }

  static reassignErrors(errors: any[]) {
    if (!Array.isArray(errors)) {
      return errors;
    }
    for (let error of errors) {
      // missing properties are propagated into properties themselves
      // rendering as "value is required"
      if (error.params && error.params.missingProperty) {
        error.dataPath += '.' + error.params.missingProperty;
        error.message = 'Value is required';
      }
    }
    return errors;
  }

  static parseParent(dataPath: string) {
    if (dataPath.indexOf('.') >= 0) {
      return {
        dataPath: dataPath.substring(0, dataPath.lastIndexOf('.')),
        property: dataPath.substring(dataPath.lastIndexOf('.') + 1)
      };
    }
    return { property: dataPath, dataPath: '' };
  }

  tryParse(value: any) {
    switch (this.type) {
      case 'date':
        let date = Date.parse(value);
        return isNaN(date) ? value : new Date(date);
      case 'integer':
        return validations.IntValidator(value) ? value : parseInt(value, 10);
      case 'number':
        return validations.FloatValidator(value) ? value : parseFloat(value);
      case 'boolean':
        return value === true || value === 'true' || value === 'True'
          ? true
          : !value || value === false || value === 'false' || value === 'False'
          ? false
          : value;
      default:
        return value;
    }
  }

  assignErrors(dataset: DataSet, error: Ajv.ErrorObject) {
    let { property, dataPath } = Schema.parseParent(error.dataPath);
    const node = eval(`dataset${dataPath}`);

    if (!property) {
      property = 'ROOT';
    }

    if (!node.errors.get(property)) {
      node.errors.set(property, error.message);
    }
  }

  validate(dataset: DataSet) {
    const cleanData = dataset.toJS ? dataset.toJS() : dataset;
    // console.log(cleanData);

    // we use this when executing expressions
    // (this.ajv as any).currentData = cleanData;

    if (!this.validator(cleanData) as any) {
      return Schema.reassignErrors(this.validator.errors);
    }
    return false;
  }

  validateAndAssignErrors(dataset: DataSet) {
    const cleanData = dataset.toJS();
    const errors = this.validate(cleanData);
    if (errors && errors.length) {
      for (let error of errors) {
        this.assignErrors(dataset, error);
      }
    }
    return errors;
  }

  validateField(value: IAnyStateTreeNode, key: string, assignErrors = false) {
    // mobx's 'getPath' generates path such as /prop/path/0
    // we need to convert it to the json schema type path such as.prop.path[0]
    const path = Schema.convertPath(getPath(value)) + (key ? '.' + key : '');
    // console.log('Validating: ' + path);

    // validate the whole dataset from its root
    const root: DataSet = getRoot(value);
    const errors = this.validate(root as DataSet);

    // console.log(root.fatherAge)
    // console.log(errors);

    // locate the error in the resulting errors by its path
    if (errors && errors.length) {
      const error = errors.find(e => e.dataPath === path);
      if (error) {
        if (assignErrors) {
          this.assignErrors(root, error);
        }
        return error.message;
      }
    }
    return undefined;
  }
}
