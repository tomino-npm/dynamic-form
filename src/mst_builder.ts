import { types } from 'mobx-state-tree';
import { UndoManager } from 'mst-middlewares';

import { Schema } from './data_schema_model';
import { FormStore } from './form_store';
import { safeEval } from './form_utils';

let time = Date.now();
let i = 0;
function shortId() {
  return (time + i++).toString();
}

function mstTypeFactory(desc: Schema): any {
  switch (desc.type) {
    case 'id':
      return types.optional(types.identifier, shortId);
    case 'expression':
      return undefined;
    case 'array':
      return types.optional(
        types.array(types.optional(mstTypeFactory(desc.items), desc.items.defaultValue)),
        desc.default || []
      );
    case 'string':
      return types.optional(types.union(types.string, types.undefined), desc.default || '');
    case 'integer':
      return types.optional(
        types.union(types.number, types.string, types.undefined),
        desc.default || ''
      );
    case 'number':
      return types.optional(
        types.union(types.number, types.string, types.undefined),
        desc.default || ''
      );
    case 'date':
      return types.optional(
        types.union(types.Date, types.string, types.undefined),
        desc.default || ''
      );
    case 'boolean':
      return types.optional(
        types.union(types.boolean, types.string, types.undefined),
        desc.default || ''
      );
    case 'object':
      return types.optional(buildStore(desc), {});
    case undefined:
      return types.string;
  }
  throw new Error('MST Type not supported: ' + desc.type);
}

export function buildStore(schema: Schema) {
  // prepare model and views

  const mstDefinition: { [index: string]: any } = {};

  /* =========================================================
    EXPRESSIONS
    ======================================================== */
  if (!schema.properties) {
    throw new Error('Schema does not contain any properties: ' + JSON.stringify(schema));
  }

  const properties = Object.getOwnPropertyNames(schema.properties);

  const viewDefinition = () => {
    const view = {};

    for (let key of properties) {
      let node = schema.properties[key];
      // expressions do not need state tree entry they are evaluated automatically
      if (node.type === 'expression') {
        (view as any).__defineGetter__(key, function() {
          // @ts-ignore
          return safeEval(this, node.default);
        });
      }
    }

    return view;
  };

  /* =========================================================
      MST Nodes
     ======================================================== */

  for (let key of properties) {
    let node = schema.properties[key];
    let definition = mstTypeFactory(node);
    if (definition) {
      mstDefinition[key] = types.maybe(definition);
    }
  }

  if (schema.parent == null) {
    mstDefinition.history = types.optional(UndoManager, {});
  }

  // build tree
  const mst = FormStore.named('FormStore')
    .props(mstDefinition)
    .views(viewDefinition)
    .actions(() => ({
      getSchema(key: string) {
        if (!key) {
          return schema;
        }
        if (key.indexOf('.') >= 0) {
          const parts = key.split('.');
          let property = schema;
          do {
            const first = parts.shift();
            property = property.properties[first];
            if (!property) {
              throw new Error(
                `Could not find key ${first} for key ${key} in schema with properties [${Object.getOwnPropertyNames(
                  schema.properties
                ).join(',')}]`
              );
            }
          } while (parts.length > 0);
          return property;
        }

        const value = key ? schema.properties[key] : schema;
        if (!value) {
          throw new Error(
            `Could not find key ${key} in schema with properties [${Object.getOwnPropertyNames(
              schema.properties
            ).join(',')}]`
          );
        }
        return value;
      }
    }));

  return mst;
}
