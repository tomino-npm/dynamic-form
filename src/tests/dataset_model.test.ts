import { autorun, toJS } from 'mobx';

import { Schema } from '../data_schema_model';
import { buildStore } from '../mst_builder';
import { create } from './data';
import { JSONSchema } from '../json_schema';

import { config } from '../config';
import { safeEval } from '../form_utils';

describe('Dataset', () => {
  let jsonSchema = (): JSONSchema => ({
    type: 'object',
    properties: {
      name: {
        type: 'string'
      },
      age: {
        type: 'integer',
        minimum: 0,
        maximum: 130,
        default: 0
      },
      salary: {
        type: 'number',
        minimum: 0,
        default: 300
      },
      married: {
        type: 'boolean',
        default: false
      },
      dateMarried: {
        type: 'string'
      },
      fatherAge: {
        type: 'integer',
        validationExpression: 'value > this.age + 18',
        validationMessage: 'Father age must be at least 18 years more then your age'
      },
      accounts: {
        type: 'array',
        uniqueItems: true,
        items: {
          type: 'object',
          properties: {
            number: {
              type: 'string',
              pattern: '\\d\\d\\d-\\d\\d\\d'
            },
            money: {
              type: 'number',
              validationExpression: 'this.money % 2 === 0', // allow only to add higher value,
              validationMessage: 'You can only put even value of money!'
            }
          },
          required: ['number']
        },

        minItems: 1,
        maxItems: 3
      },
      accountTotal: {
        type: 'number',
        expression: 'this.accounts.reduce((p, n) => n.money + p, 0)'
      },
      address: {
        type: 'object',
        properties: {
          street: {
            type: 'string',
            minLength: 5,
            maxLength: 30
          },
          number: {
            type: 'number',
            exclusiveMinimum: 0,
            exclusiveMaximum: 1000
          }
        }
      }
    },
    required: ['name']
  });

  let schema = new Schema(jsonSchema());

  it('creates a default value', () => {
    expect(schema.defaultValue()).toMatchSnapshot();

    expect(schema.properties.accounts.items.defaultValue()).toMatchSnapshot();
  });

  it('creates a new mst and validates values', () => {
    config.setDirty = jest.fn();

    const mst = buildStore(schema);
    const data = mst.create({});
    let error = '';
    // check conversion

    /* =========================================================
        Integer / Number
       ======================================================== */

    // ok

    data.setValue('age', '30');
    expect(data.getValue('age')).toEqual(30);

    data.setValue('salary', '30.15');
    expect(data.getValue('salary')).toEqual(30.15);

    data.setValue('salary', '30');
    expect(data.getValue('salary')).toEqual(30);

    // // format

    data.setValue('age', '30.15');
    expect(data.getValue('age')).toEqual('30.15');
    expect(data.validateField('age')).toBe('should be integer');

    data.setValue('age', 'Momo');
    expect(data.getValue('age')).toEqual('Momo');
    expect(data.validateField('age')).toBe('should be integer');

    data.setValue('salary', '30.aa');
    expect(data.getValue('salary')).toEqual('30.aa');
    expect(data.validateField('salary')).toBe('should be number');

    // minimum

    data.setValue('age', '0');
    expect(data.getValue('age')).toEqual(0);

    data.setValue('age', '-1');
    expect(data.getValue('age')).toEqual(-1);
    expect(data.validateField('age')).toBe('should be >= 0');

    // // maximum

    data.setValue('age', '130');
    expect(data.getValue('age')).toEqual(130);

    data.setValue('age', '131');
    expect(data.getValue('age')).toEqual(131);
    expect(data.validateField('age')).toBe('should be <= 130');

    // subselection ok 'a.b.c'

    data.setValue('address.number', '20');
    expect(data.getValue('address.number')).toEqual(20);

    // exclusive minimum

    data.setValue('address.number', '0');
    expect(data.getValue('address.number')).toEqual(0);
    expect(data.getValue('address').validateField('number')).toBe('should be > 0');

    data.setValue('address.number', '1');
    expect(data.getValue('address.number')).toEqual(1);

    // exclusive maximum

    data.setValue('address.number', '1000');
    expect(data.getValue('address.number')).toEqual(1000);
    expect(data.getValue('address').validateField('number', '1000')).toBe('should be < 1000');

    data.setValue('address.number', '999');
    expect(data.getValue('address.number')).toEqual(999);

    /* =========================================================
        String
       ======================================================== */

    // ok

    data.setValue('name', 'Bobo');
    expect(data.getValue('name')).toEqual('Bobo');

    // format ok

    data.addRow('accounts');
    data.getValue('accounts')[0].setValue('number', '234-234');

    // console.log(3);
    // console.log(data.getValue('accounts')[0]);

    expect(data.getValue('accounts')[0].number).toEqual('234-234');

    error = data.getValue('accounts')[0].validateField('number', '123-123');
    expect(error).toBeUndefined();

    // format error

    data.getValue('accounts')[0].setValue('number', '123d');
    expect(data.getValue('accounts')[0].number).toEqual('123d');
    error = data.getValue('accounts')[0].validateField('number');
    expect(error).toEqual('should match pattern "\\d\\d\\d-\\d\\d\\d"');

    data.getValue('accounts')[0].setValue('number', '');
    error = data.getValue('accounts')[0].validateField('number', '');
    expect(error).toEqual('Value is required');

    // minLength

    data.getValue('address').setValue('street', '123456');
    error = data.getValue('address').validateField('street', '123456');
    expect(error).toBeUndefined();

    data.getValue('address').setValue('street', '123');
    error = data.getValue('address').validateField('street');
    expect(error).toEqual('should NOT be shorter than 5 characters');

    // maxLength

    data.getValue('address').setValue('street', '1234567890123456789012345678901234567890');
    error = data.getValue('address').validateField('street');
    expect(error).toEqual('should NOT be longer than 30 characters');

    /* =========================================================
        Array
       ======================================================== */

    // min intems

    data.removeRow('accounts', 0);
    error = data.validateField('accounts');
    expect(error).toEqual('should NOT have fewer than 1 items');

    // max items

    data.addRow('accounts');
    data.addRow('accounts');
    data.addRow('accounts');
    data.addRow('accounts');

    error = data.validateField('accounts');
    expect(error).toEqual('should NOT have more than 3 items');

    data.removeRow('accounts', 3);

    // unique items

    // console.log(toJS(data.getValue('accounts')));
    // console.log(data.getValue('accounts'))

    error = data.validateField('accounts');
    expect(error).toBe('should NOT have duplicate items (items ## 1 and 2 are identical)');

    data.getValue('accounts')[1].setValue('number', '123-234');

    error = data.validateField('accounts');
    expect(error).toBe('should NOT have duplicate items (items ## 0 and 2 are identical)');

    // all is well

    data.getValue('accounts')[0].setValue('number', '123-456');
    data.getValue('accounts')[2].setValue('number', '456-789');

    error = data.validateField('accounts');
    expect(error).toBeUndefined();

    // unique items

    /* =========================================================
        Expression
       ======================================================== */

    data.setValue('age', '10');
    data.setValue('fatherAge', '40');
    expect(data.getValue('fatherAge')).toEqual(40);

    // failing expression

    data.setValue('fatherAge', '20');
    expect(data.getValue('fatherAge')).toEqual(20);

    // console.log('WILL VALIDATE FATHER AGE');
    // console.log(data.fatherAge);

    error = data.validateField('fatherAge');
    expect(error).toBe('Father age must be at least 18 years more then your age');

    data.setValue('fatherAge', '80');
    expect(data.getValue('fatherAge')).toEqual(80);
    error = data.validateField('fatherAge');
    expect(error).toBeUndefined();

    data.getValue('accounts')[0].setValue('money', 10);
    data.getValue('accounts')[0].setValue('money', 5);
    error = data.getValue('accounts')[0].validateField('money');
    expect(error).toEqual('You can only put even value of money!');
    expect(data.getValue('accounts')[0].errors.get('money')).toBe(
      'You can only put even value of money!'
    );

    data.getSchema('fatherAge').validationMessage = null;
  });

  it('creates mst with values', () => {
    const mst = buildStore(schema);
    const data = mst.create({
      age: 50,
      name: 'Tomas',
      fatherAge: 20,
      married: true,
      dateMarried: create.date().toISOString(),
      salary: 2300.34,
      address: {
        street: 'Elm street',
        number: 4
      },
      accounts: [{ number: '111-222', money: 200 }]
    });

    const m = 1;
    // console.log(create.date new Date(2012, 8, 29, 0, 0, 0, 0).toISOString())

    expect(data.toJS()).toMatchSnapshot();
  });

  it('creates mst with default values', () => {
    const mst = buildStore(schema);
    const defaultData = mst.create({});
    expect(defaultData.toJS(false)).toMatchSnapshot();
  });

  it('validates the root dataset', () => {
    const jSchema = jsonSchema();
    jSchema.oneOf = [{ required: ['name'] }, { required: ['age'] }];
    const rSchema = new Schema(jSchema);

    const store = buildStore(rSchema);
    const d1 = store.create({ name: 'Tomas', age: 39 });
    const errors = d1.validateDataset(); /*?*/
    expect(d1.errors.get('ROOT')).toEqual('should match exactly one schema in oneOf');
  });

  it('allows to use expressions', () => {
    const mst = buildStore(schema);
    const defaultData = mst.create({
      accounts: [{ number: '1', money: 1 }, { number: '2', money: 2 }, { number: '3', money: 3 }]
    });

    expect(defaultData.getValue('accountTotal')).toBe(6);

    let finalAccount = 0;
    autorun(() => {
      finalAccount = defaultData.getValue('accountTotal');
    });

    // check computed fields
    defaultData.getValue('accounts')[0].setValue('money', 10);
    expect(finalAccount).toEqual(15);
  });
});
