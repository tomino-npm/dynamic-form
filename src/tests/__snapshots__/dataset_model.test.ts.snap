// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dataset creates a default value 1`] = `
Object {
  "age": 0,
  "married": false,
  "salary": 300,
}
`;

exports[`Dataset creates a default value 2`] = `Object {}`;

exports[`Dataset creates mst with default values 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset creates mst with default values 2`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset creates mst with values 1`] = `
Object {
  "accounts": Array [
    Object {
      "money": 200,
      "number": "111-222",
    },
  ],
  "address": Object {
    "number": 4,
    "street": "Elm street",
  },
  "age": 50,
  "dateMarried": "2017-02-02T00:00:00.000Z",
  "fatherAge": 20,
  "married": true,
  "name": "Tomas",
  "salary": 2300.34,
}
`;

exports[`Dataset validates  1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset validates s 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset validates th 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset validates the  1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset validates the  dataset 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset validates the d 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset validates the da 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset validates the datas 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset validates the dataset 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset validates the r dataset 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset validates the root dataset 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset validatess 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset validats 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset valids 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;

exports[`Dataset vs 1`] = `
Object {
  "accounts": Array [],
  "address": Object {
    "number": "",
    "street": "",
  },
  "age": "",
  "dateMarried": "",
  "fatherAge": "",
  "married": "",
  "name": "",
  "salary": 300,
}
`;
